### PR TITLE
Upgrade to clang-tidy 18

### PR DESCRIPTION
**Pull request by SiarheiFedartsou**
<details><summary><h2>Benchmark Results</h2></summary>

| Benchmark | Base | PR |
|-----------|------|----|
| alias | aliased u32: 1223.27<br/>plain u32: 1213.06<br/>aliased double: 1279.07<br/>plain double: 1272.1 | aliased u32: 1180.83<br/>plain u32: 1170.75<br/>aliased double: 1236.3<br/>plain double: 1224.54 |
| e2e_match_ch | requests: 381<br/>failures: 0<br/>req/s: 6.352req/s<br/>avg: 5.031ms<br/>50%: 3ms<br/>75%: 4ms<br/>95%: 13ms<br/>98%: 30ms<br/>99%: 45ms   <br/>min: 1.318ms<br/>max: 78.540ms | requests: 362<br/>failures: 0<br/>req/s: 6.038req/s<br/>avg: 5.257ms<br/>50%: 3ms<br/>75%: 4ms<br/>95%: 11ms<br/>98%: 33ms<br/>99%: 63ms   <br/>min: 1.262ms<br/>max: 76.613ms |
| e2e_match_mld | requests: 342<br/>failures: 0<br/>req/s: 5.705req/s<br/>avg: 5.362ms<br/>50%: 4ms<br/>75%: 4ms<br/>95%: 15ms<br/>98%: 30ms<br/>99%: 42ms   <br/>min: 1.353ms<br/>max: 60.721ms | requests: 379<br/>failures: 0<br/>req/s: 6.317req/s<br/>avg: 5.328ms<br/>50%: 3ms<br/>75%: 4ms<br/>95%: 15ms<br/>98%: 28ms<br/>99%: 59ms   <br/>min: 1.362ms<br/>max: 97.454ms |
| e2e_nearest_ch | requests: 438<br/>failures: 0<br/>req/s: 7.303req/s<br/>avg: 1.697ms<br/>50%: 2ms<br/>75%: 2ms<br/>95%: 2ms<br/>98%: 2ms<br/>99%: 2ms   <br/>min: 1.169ms<br/>max: 4.689ms | requests: 411<br/>failures: 0<br/>req/s: 6.855req/s<br/>avg: 1.624ms<br/>50%: 2ms<br/>75%: 2ms<br/>95%: 2ms<br/>98%: 3ms<br/>99%: 3ms   <br/>min: 1.066ms<br/>max: 6.116ms |
| e2e_nearest_mld | requests: 435<br/>failures: 0<br/>req/s: 7.257req/s<br/>avg: 1.897ms<br/>50%: 2ms<br/>75%: 2ms<br/>95%: 2ms<br/>98%: 3ms<br/>99%: 3ms   <br/>min: 1.219ms<br/>max: 52.939ms | requests: 408<br/>failures: 0<br/>req/s: 6.801req/s<br/>avg: 1.605ms<br/>50%: 2ms<br/>75%: 2ms<br/>95%: 2ms<br/>98%: 2ms<br/>99%: 2ms   <br/>min: 1.032ms<br/>max: 6.054ms |
| e2e_route_ch | requests: 382<br/>failures: 0<br/>req/s: 6.369req/s<br/>avg: 4.755ms<br/>50%: 4ms<br/>75%: 5ms<br/>95%: 7ms<br/>98%: 9ms<br/>99%: 10ms   <br/>min: 1.467ms<br/>max: 101.833ms | requests: 378<br/>failures: 0<br/>req/s: 6.304req/s<br/>avg: 4.746ms<br/>50%: 4ms<br/>75%: 5ms<br/>95%: 6ms<br/>98%: 8ms<br/>99%: 10ms   <br/>min: 1.476ms<br/>max: 143.283ms |
| e2e_route_mld | requests: 384<br/>failures: 0<br/>req/s: 6.406req/s<br/>avg: 5.409ms<br/>50%: 5ms<br/>75%: 6ms<br/>95%: 7ms<br/>98%: 10ms<br/>99%: 11ms   <br/>min: 1.386ms<br/>max: 99.600ms | requests: 412<br/>failures: 0<br/>req/s: 6.868req/s<br/>avg: 4.547ms<br/>50%: 4ms<br/>75%: 5ms<br/>95%: 6ms<br/>98%: 8ms<br/>99%: 10ms   <br/>min: 1.610ms<br/>max: 98.198ms |
| e2e_table_ch | requests: 376<br/>failures: 0<br/>req/s: 6.269req/s<br/>avg: 49.679ms<br/>50%: 30ms<br/>75%: 87ms<br/>95%: 130ms<br/>98%: 140ms<br/>99%: 140ms   <br/>min: 3.030ms<br/>max: 274.284ms | requests: 392<br/>failures: 0<br/>req/s: 6.538req/s<br/>avg: 56.104ms<br/>50%: 38ms<br/>75%: 94ms<br/>95%: 130ms<br/>98%: 140ms<br/>99%: 150ms   <br/>min: 2.608ms<br/>max: 332.286ms |
| e2e_table_mld | requests: 371<br/>failures: 0<br/>req/s: 6.189req/s<br/>avg: 75.046ms<br/>50%: 77ms<br/>75%: 110ms<br/>95%: 140ms<br/>98%: 150ms<br/>99%: 150ms   <br/>min: 4.686ms<br/>max: 183.313ms | requests: 414<br/>failures: 0<br/>req/s: 6.901req/s<br/>avg: 45.394ms<br/>50%: 28ms<br/>75%: 75ms<br/>95%: 130ms<br/>98%: 140ms<br/>99%: 150ms   <br/>min: 2.660ms<br/>max: 159.087ms |
| e2e_trip_ch | requests: 363<br/>failures: 0<br/>req/s: 6.052req/s<br/>avg: 18.241ms<br/>50%: 17ms<br/>75%: 24ms<br/>95%: 32ms<br/>98%: 34ms<br/>99%: 36ms   <br/>min: 2.449ms<br/>max: 41.356ms | requests: 341<br/>failures: 0<br/>req/s: 5.687req/s<br/>avg: 18.049ms<br/>50%: 17ms<br/>75%: 24ms<br/>95%: 32ms<br/>98%: 33ms<br/>99%: 34ms   <br/>min: 2.060ms<br/>max: 37.663ms |
| e2e_trip_mld | requests: 354<br/>failures: 0<br/>req/s: 5.906req/s<br/>avg: 21.897ms<br/>50%: 21ms<br/>75%: 28ms<br/>95%: 35ms<br/>98%: 42ms<br/>99%: 44ms   <br/>min: 4.209ms<br/>max: 46.124ms | requests: 354<br/>failures: 0<br/>req/s: 5.901req/s<br/>avg: 16.214ms<br/>50%: 15ms<br/>75%: 21ms<br/>95%: 31ms<br/>98%: 33ms<br/>99%: 35ms   <br/>min: 2.675ms<br/>max: 44.021ms |
| json-render | String: 6.75339ms<br/>Stringstream: 9.4243ms<br/>Vector: 6.98778ms | String: 6.70346ms<br/>Stringstream: 9.32243ms<br/>Vector: 6.91688ms |
| match_ch | Default radius: <br/>4.41954ms/req at 82 coordinate<br/>0.0538968ms/coordinate<br/>Radius 5m: <br/>4.38232ms/req at 82 coordinate<br/>0.0534429ms/coordinate<br/>Radius 10m: <br/>14.9263ms/req at 82 coordinate<br/>0.182028ms/coordinate<br/>Radius 15m: <br/>36.6036ms/req at 82 coordinate<br/>0.446385ms/coordinate<br/>Radius 30m: <br/>312.995ms/req at 82 coordinate<br/>3.81701ms/coordinate | Default radius: <br/>4.45352ms/req at 82 coordinate<br/>0.0543112ms/coordinate<br/>Radius 5m: <br/>4.43942ms/req at 82 coordinate<br/>0.0541393ms/coordinate<br/>Radius 10m: <br/>15.1616ms/req at 82 coordinate<br/>0.184898ms/coordinate<br/>Radius 15m: <br/>37.1285ms/req at 82 coordinate<br/>0.452786ms/coordinate<br/>Radius 30m: <br/>315.59ms/req at 82 coordinate<br/>3.84866ms/coordinate |
| match_mld | Default radius: <br/>2.76855ms/req at 82 coordinate<br/>0.0337628ms/coordinate<br/>Radius 5m: <br/>2.83892ms/req at 82 coordinate<br/>0.0346209ms/coordinate<br/>Radius 10m: <br/>10.1743ms/req at 82 coordinate<br/>0.124077ms/coordinate<br/>Radius 15m: <br/>26.1871ms/req at 82 coordinate<br/>0.319354ms/coordinate<br/>Radius 30m: <br/>302.946ms/req at 82 coordinate<br/>3.69446ms/coordinate | Default radius: <br/>2.75393ms/req at 82 coordinate<br/>0.0335845ms/coordinate<br/>Radius 5m: <br/>2.73113ms/req at 82 coordinate<br/>0.0333065ms/coordinate<br/>Radius 10m: <br/>10.0511ms/req at 82 coordinate<br/>0.122574ms/coordinate<br/>Radius 15m: <br/>25.7732ms/req at 82 coordinate<br/>0.314308ms/coordinate<br/>Radius 30m: <br/>302.781ms/req at 82 coordinate<br/>3.69245ms/coordinate |
| packedvector | random write:<br/>std::vector 12147.1 ms<br/>util::packed_vector 81904.8 ms<br/>slowdown: 6.74273<br/>random read:<br/>std::vector 12018.3 ms<br/>util::packed_vector 33614.6 ms<br/>slowdown: 2.79696 | random write:<br/>std::vector 11579.4 ms<br/>util::packed_vector 73798.2 ms<br/>slowdown: 6.37324<br/>random read:<br/>std::vector 11333.3 ms<br/>util::packed_vector 30349.3 ms<br/>slowdown: 2.6779 |
| route_ch | 1000 routes, 3 coordinates, no alternatives, overview=full, steps=true<br/>514.012ms<br/>0.514012ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=full, steps=true<br/>351.655ms<br/>0.351655ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=full, steps=true<br/>629.65ms<br/>0.62965ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false<br/>151.993ms<br/>0.151993ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false<br/>97.6789ms<br/>0.0976789ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false<br/>133.34ms<br/>0.13334ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>152.201ms<br/>0.152201ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>98.2297ms<br/>0.0982297ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false, radius=750<br/>132.641ms<br/>0.132641ms/req | 1000 routes, 3 coordinates, no alternatives, overview=full, steps=true<br/>511.213ms<br/>0.511213ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=full, steps=true<br/>353.241ms<br/>0.353241ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=full, steps=true<br/>667.345ms<br/>0.667345ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false<br/>150.854ms<br/>0.150854ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false<br/>97.4697ms<br/>0.0974697ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false<br/>133.423ms<br/>0.133423ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>151.391ms<br/>0.151391ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>97.2257ms<br/>0.0972257ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false, radius=750<br/>131.898ms<br/>0.131898ms/req |
| route_mld | 1000 routes, 3 coordinates, no alternatives, overview=full, steps=true<br/>639.617ms<br/>0.639617ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=full, steps=true<br/>439.342ms<br/>0.439342ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=full, steps=true<br/>814.715ms<br/>0.814715ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false<br/>263.551ms<br/>0.263551ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false<br/>160.555ms<br/>0.160555ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false<br/>283.312ms<br/>0.283312ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>259.53ms<br/>0.25953ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>159.99ms<br/>0.15999ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false, radius=750<br/>282.618ms<br/>0.282618ms/req | 1000 routes, 3 coordinates, no alternatives, overview=full, steps=true<br/>646.409ms<br/>0.646409ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=full, steps=true<br/>441.817ms<br/>0.441817ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=full, steps=true<br/>820.048ms<br/>0.820048ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false<br/>266.992ms<br/>0.266992ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false<br/>162.773ms<br/>0.162773ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false<br/>285.496ms<br/>0.285496ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>258.837ms<br/>0.258837ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>160.161ms<br/>0.160161ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false, radius=750<br/>286.399ms<br/>0.286399ms/req |
| rtree | 1 result:<br/>206.729ms ->  0.0206729 ms/query<br/>10 results:<br/>242.198ms ->  0.0242198 ms/query | 1 result:<br/>207.879ms ->  0.0207879 ms/query<br/>10 results:<br/>242.733ms ->  0.0242733 ms/query |
</details>

### CODE REVIEW

File: .clang-tidy [hunk 1]
```diff
@@ -13,6 +13,11 @@ Checks: >
     -bugprone-forward-declaration-namespace,
     -bugprone-sizeof-expression,
     -bugprone-throw-keyword-missing,
+    -bugprone-chained-comparison,
```
Comment (SiarheiFedartsou) [inline .clang-tidy:16]: Many of these warnings seem to be useful - I'll create an issue to fix some of them before merging it.

File: third_party/sol2/include/sol/sol.hpp [hunk 2]
```diff
@@ -19416,7 +19416,13 @@ namespace sol { namespace function_detail {
 		}
 
 		template <bool is_yielding, bool no_trampoline>
-		static int call(lua_State* L) noexcept(std::is_nothrow_copy_assignable_v<T>) {
+		static int call(lua_State* L) 
+#if SOL_IS_ON(SOL_COMPILER_CLANG)
```
Comment (SiarheiFedartsou) [inline third_party/sol2/include/sol/sol.hpp:19420]: Took this workaround from https://github.com/ThePhD/sol2/issues/1581#issuecomment-2103463524

I don't see another way to build with newer clang unfortunately. I'll try to make a PR to have this workaround in sol2 mainline...

File: .github/workflows/osrm-backend.yml [hunk 3]
```diff
@@ -703,7 +703,7 @@ jobs:
         run: |
           ccache -p
           ccache -s
-
+
```
Comment (DennisOSRM) [inline .github/workflows/osrm-backend.yml:706]: nit pick: stray whitespace 😉

File: third_party/sol2/include/sol/sol.hpp (hunk 2 above)
Comment (DennisOSRM) [reply third_party/sol2/include/sol/sol.hpp:19420]: Consider adding a link as a comment right in the source. This way one can trace the origin of this bandaid, and eventually remove it in the future.

File: unit_tests/library/route.cpp [hunk 4]
```diff
@@ -239,7 +239,6 @@ void test_route_same_coordinates(bool use_json_only_api)
                 BOOST_CHECK(((void)name, true));
 
                 // nothing can be said about mode, contains mode of transportation
-                const auto mode = std::get<json::String>(step_object.values.at("mode")).value;
```
Comment (DennisOSRM) [inline unit_tests/library/route.cpp:242]: The comment above looks like it can be removed, too.

**Review by DennisOSRM — APPROVED** [review]
Nice to see the toolchain get updated. 👍🏽

File: .clang-tidy (hunk 1 above)
Comment (SiarheiFedartsou) [reply .clang-tidy:16]: https://github.com/Project-OSRM/osrm-backend/issues/6928

**Pull request merged**